File: ChessEngine/queen.py
rook_offsets = [(1, 0), (-1, 0), (0, 1), (0, -1)]
bishop_offsets = [(1, 1), (1, -1), (-1, 1), (-1, -1)]
queen_offsets = rook_offsets + bishop_offsets

def check_queen(current_board: dict,moves_list: list ,value="queen"):
    """
    The function `check_queen` iterates through a chess board to find possible moves for queens and
    updates a list of moves accordingly.
    
    :param current_board: The function `check_queen` seems to be checking for possible moves of a queen
    on a chessboard based on the current board state. It iterates over the positions on the board and
    checks for possible moves in all directions for both white and black queens
    :type current_board: dict
    :param moves_list: The `moves_list` parameter is a dictionary that contains lists of moves for
    different pieces. In the function `check_queen`, it seems to be used to store the possible moves for
    the queen pieces on the chessboard. The keys `'b'` and `'w'` likely represent black and
    :type moves_list: list
    :param value: The `value` parameter in the `check_queen` function is used to specify the type of
    chess piece to check for. By default, it is set to "queen". This parameter allows the function to
    differentiate between checking for a white queen ("Q") and a black queen ("q") on, defaults to queen
    (optional)
    """
    chain = 1 
    check = True
    
    for k, v in current_board.items():
        check = True
        if v == "Q":
            for offset in queen_offsets:
                check = True
                chain = 1
                while check:
                    if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and (current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].islower() or current_board[(k[0] + \
                            chain * offset[0], k[1] + chain * offset[1])] == ""):
                        
                        chain += 1
                        
                        if check:
                            moves_list['b'].append((k, (k[0] + (chain * offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1]))))
                            
                        if (k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1]) in current_board and current_board[(k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1])].islower():
                            
                            check = False
                            chain += 1
                            moves_list['b'].append((k, (k[0] + (chain * offset[0] - offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1] - offset[1]))))
                    else:
                        check = False
        if v == "q":
            for offset in queen_offsets:
                check = True
                chain = 1
                while check:
                    if (k[0] + chain * offset[0], k[1] + chain * offset[1]) in current_board and (current_board[(k[0] + chain * offset[0], k[1] + chain * offset[1])].isupper() or current_board[(k[0] + \
                            chain * offset[0], k[1] + chain * offset[1])] == ""):
                        
                        chain += 1
                        
                        if check:
                            moves_list['w'].append((k, (k[0] + (chain * offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1]))))
                            
                        if (k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1]) in current_board and current_board[(k[0] + chain * offset[0] - offset[0], k[1] + chain * offset[1] - offset[1])].isupper():
                            
                            check = False
                            chain += 1
                            moves_list['w'].append((k, (k[0] + (chain * offset[0] - offset[0] - offset[0]), k[1] + (chain * offset[1] - offset[1] - offset[1]))))
                    else:
                        check = False

File: ChessEngine/chessEngineCPP/queen.cpp
#include "queen.h"
#include <map>
#include <vector>

// Queen move offsets (rook + bishop)
const std::pair<int, int> queen_offsets[8] = {
    {1, 0}, {-1, 0}, {0, 1}, {0, -1}, // Rook-like moves
    {1, 1}, {1, -1}, {-1, 1}, {-1, -1} // Bishop-like moves
};

void check_queen(const std::map<std::pair<int, int>, char>& current_board, 
                  std::map<char, std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>>>& moves_list) 
{
    for (const auto& piece : current_board) {
        std::pair<int, int> position = piece.first;
        char piece_type = piece.second;
        int chain = 1;
        
        // Check for white and black bishops
        if (piece_type == 'Q' || piece_type == 'q') {
            for (const auto& offset : queen_offsets) {
                chain = 1;
                bool check = true;
                
                while (check) {
                    std::pair<int, int> target_position = {
                        position.first + chain * offset.first,
                        position.second + chain * offset.second
                    };
                    
                    // If target position is on the board
                    if (current_board.find(target_position) != current_board.end()) {
                        char target_piece = current_board.at(target_position);
                        if ((piece_type == 'Q' && isupper(target_piece)) ||
                            (piece_type == 'q' && islower(target_piece))) {
                            check = false;
                            break;
                        }
                        // Determine if it's a valid move for black or white
                        if (piece_type == 'Q' && (target_piece == ' ' || islower(target_piece))) {
                            moves_list['b'].push_back({position, target_position});
                        } else if (piece_type == 'q' && (target_piece == ' ' || isupper(target_piece))) {
                            moves_list['w'].push_back({position, target_position});
                        }
                        
                        // Stop if there's an enemy piece to capture
                        if ((piece_type == 'Q' && islower(target_piece)) ||
                            (piece_type == 'q' && isupper(target_piece))) {
                            check = false;
                        }
                        
                    chain++;
                    } else {
                        check = false;
                    }
                }
            }
        }
    }
}
